client/js/widgets/widget.js

```diff
@@ -200,10 +241,543 @@ export class Widget extends StateManaged {
+      }
+
+    }
+
+    showOverlay(null);
+    if(go)
+      resolve(result);
+    else
+      reject(result);
+  }
+
+  async evaluateRoutine(property, initialVariables, initialCollections, depth) {
+    function setDefaults(routine, defaults) {
+      for(const key in defaults)
+        if(routine[key] === undefined)
+          routine[key] = defaults[key];
+    }
+
+    function isValidCollection(collection) {
+      if(Array.isArray(collections[collection]))
+        return true;
+      problems.push(`Collection ${collection} does not exist.`);
+    }
+
+    function toA(ids) {
+      return typeof ids == 'string' ? [ ids ] : ids;
+    }
+
+    function w(ids, callback) {
+      return widgetFilter(w=>toA(ids).indexOf(w.p('id')) != -1).forEach(callback);
+    }
+
+    batchStart();
+
+    if(this.p('debug') && !depth)
+      $('#debugButtonOutput').textContent = '';
+
+    const variables = Object.assign({
+      playerName,
+      playerColor,
+      activePlayers,
+      thisID : this.p('id')
+    }, initialVariables);
+
+    const collections = Object.assign({
+      thisButton : [this]
+    }, initialCollections);
+
+    for(const original of this.p(property)) {
+      const a = JSON.parse(JSON.stringify(original));
+      var problems = [];
+
+      if(this.p('debug')) console.log(`${this.id}: ${JSON.stringify(original)}`);
+
+      if(a.applyVariables) this.applyVariables(a, variables, problems);
+
+      if(a.skip) {
+        $('#debugButtonOutput').textContent += '\n\n\nOPERATION SKIPPED: \n' + JSON.stringify(a, null, '  ');
+        continue;
+      }
+
+      if(a.func == 'CALL') {
+        setDefaults(a, { widget: this.p('id'), routine: 'clickRoutine', 'return': true, arguments: {}, variable: 'result' });
+        if(!a.routine.match(/Routine$/)) {
+          problems.push('Routine parameters have to end with "Routine".');
+        } else if(this.isValidID(a.widget)) {
+          if(!Array.isArray(widgets.get(a.widget).p(a.routine))) {
+            problems.push(`Widget ${a.widget} does not contain ${a.routine} (or it is no array).`);
+          } else {
+            // make sure everything is passed in a way that the variables and collections of this routine won't be changed
+            const inheritVariables = Object.assign(JSON.parse(JSON.stringify(variables)), a.arguments);
+            const inheritCollections = {};
+            for(const c in collections)
+              inheritCollections[c] = [ ...collections[c] ];
+            inheritCollections['caller'] = [ this ];
+            $('#debugButtonOutput').textContent += `\n\n\nCALLing: ${a.widget}.${a.routine}\n`;
+            variables[a.variable] = await widgets.get(a.widget).evaluateRoutine(a.routine, inheritVariables, inheritCollections, (depth || 0) + 1);
+          }
+        }
+        if(!a.return) {
+          $('#debugButtonOutput').textContent += '\n\n\nCALL without return. Ending evaluation.\n';
```

Comment:
Two questions:

1. What is the purpose of this line? It's evaluated in the context of the caller if I understand it correctly, so won't it always fire?
2. What is the purpose of the `return:false` property mentioned in the summary? I don't see `return` being used anywhere else in the code.

Reply:
The `result` is what the called routine can give back to its caller.

Setting `return` to `false` forces the calling routine to end after the `CALL` operation, turning it more into a GOTO or JUMP. And this place in the code is exactly where that happens: `if(!a.return) break;`

Reply:
I see. I didn't read the spec carefully enough. This makes sense.